src/CommonRuntime/Caching.fs

```diff
@@ -35,16 +36,19 @@ let createInMemoryCache expiration =
 open System.Collections.Concurrent
 
 /// Creates a cache that uses in-memory collection
-let createInMemoryCache expiration = 
+let createInMemoryCache (expiration:TimeSpan) = 
   let dict = new ConcurrentDictionary<_, _>()
   { new ICache<_> with
       member __.Set(key, value) = 
-        dict.[key] <- (value, DateTime.UtcNow)
+        dict.AddOrUpdate(key, (value, DateTime.UtcNow), (fun _ _ -> value, DateTime.UtcNow)) |> ignore
+        async { do! Async.Sleep (expiration.Milliseconds)
+                if dict <> null then
+                    dict.TryRemove(key) |> ignore } |> Async.Start
```

Comment:
Feel free to copy or merge just the lines you need. I won't be offended of denying this PR without merge.

For me, the NuGet-version of XMLProvider was unusable to work with (yes it works in general, but it made Visual Studio so slow that I couldn't do my job, VS consumed over 2GB of memory in 10mins work, and changing the XML-file caused VS to crash every time). I can work with my current version so I'm happy with it and I just wanted to share the modifications. 

I'm not currently having time to debug and fix the actual cause of problems but I think it has somethig to do with leaking caches and the fact that TypeProviders are executed in multiple threads in VS and too much code is run just when you call a single constructed-type constructor inside other constructed-type constructors in the DOM-tree.

So if you'll do a new release that is performing better, I'll change to that but before it I will continue using the version in my fork.